test/powershell/Modules/Microsoft.PowerShell.Utility/Add-Type.Tests.ps1

```diff
@@ -26,41 +26,212 @@ Describe "Add-Type" -Tags "CI" {
+        ## If we use $TestDrive here, then Pester will try to remove them afterward and result in errors.
+        $TempPath = [System.IO.Path]::GetTempFileName()
+        if (Test-Path $TempPath) { Remove-Item -Path $TempPath -Force -Recurse }
+        New-Item -Path $TempPath -ItemType Directory -Force > $null
 
-        { [Test.AddType.BasicTest1]::Add1(1, 2) } | Should Throw
-        { [Test.AddType.BasicTest2]::Add2(3, 4) } | Should Throw
+        $outFile = Join-Path -Path $TempPath -ChildPath "assembly$guid.dll"
 
-        Add-Type -Path $codeFile1,$codeFile2
+        $code = @"
+using System.Management.Automation;
+[System.Management.Automation.Cmdlet("Get", "CompileThing$guid", ConfirmImpact = System.Management.Automation.ConfirmImpact.High, SupportsPaging = true)]
+public class AttributeTest$guid : PSCmdlet
+{
+    protected override void EndProcessing()
+
+    {
+        WriteObject("$guid");
+    }
+}
+"@
+
+        $cmdlet = "Get-CompileThing$guid"
+
+        Add-Type -TypeDefinition $code -CompileOnly -OutputAssembly $outFile -PassThru | Should BeNullOrEmpty
+        { Invoke-Expression -Command $cmdlet } | Should Throw
+
+        $testModule = Import-Module -Name $outFile -PassThru
+        Invoke-Expression -Command $cmdlet | Should BeExactly $guid
+
+        Remove-Module $testModule -Force
+    }
+
+    It "Can report C# parse and compile errors" {
+        # Add-Type show parse and compile errors and then finish with an terminationg error.
+        # We test only for '-MemberDefinition' because '-Path' uses the same code path.
+        # In the tests the error is that 'using System.Text;' is missing.
+        #
+        # Catch non-termination information error.
+        { Add-Type -MemberDefinition "public static string TestString() { return UTF8Encoding.UTF8.ToString();}" -Name "TestType1" -Namespace "TestNS" -ErrorAction Stop } | ShouldBeErrorId "SOURCE_CODE_ERROR,Microsoft.PowerShell.Commands.AddTypeCommand"
+        # Catch final terminationg error.
+        { Add-Type -MemberDefinition "public static string TestString() { return UTF8Encoding.UTF8.ToString();}" -Name "TestType1" -Namespace "TestNS" -ErrorAction SilentlyContinue } | ShouldBeErrorId "COMPILER_ERRORS,Microsoft.PowerShell.Commands.AddTypeCommand"
+
+        # Catch non-termination information error for ExtendedOptions.
+        { Add-Type -ExtendedOptions "/platform:anycpuERROR" -Language CSharp -MemberDefinition "public static string TestString() { return ""}" -Name "TestType1" -Namespace "TestNS" -ErrorAction Stop } | ShouldBeErrorId "SOURCE_CODE_ERROR,Microsoft.PowerShell.Commands.AddTypeCommand"
+        { Add-Type -ExtendedOptions "/platform:anycpuERROR" -Language VisualBasic -MemberDefinition "Public Shared Function TestString() As String `n Return `"`" `n End Function" -Name "TestType1" -Namespace "TestNS" -ErrorAction Stop } | ShouldBeErrorId "SOURCE_CODE_ERROR,Microsoft.PowerShell.Commands.AddTypeCommand"
+    }
```

Comment:
can we use the new pester builtin for checking errorid?
```
{ Add-Type -ExtendedOptions "/platform:anycpuERROR" -Language CSharp -MemberDefinition "public static string TestString() { return ""}" -Name "TestType1" -Namespace "TestNS" -ErrorAction Stop } | Should -throw -errorid "SOURCE_CODE_ERROR,Microsoft.PowerShell.Commands.AddTypeCommand"
```

Reply:
If you approve we should change the pattern in our docs https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/WritingPesterTests.md and then fix all tests.

Reply:
Currently we have a tracking Issue to fix this over all tests.